### PR TITLE
Move from TravisCI to GitHub Actions

### PR DESCRIPTION
**Pull request by desrosj**
With recent changes to the TravisCI's service, the organization is moving over to GitHub Actions. This PR will convert all of the builds in Travis to jobs in workflows within GitHub Actions.

This converts the TravisCI configuration into 4 workflow files representing the 4 different stages of testing within Travis.

I've done my best to replicate the testing stages with all of their needed parts with a few exceptions.

There is a [`workflow_run` event type](https://docs.github.com/en/free-pro-team@latest/actions/reference/events-that-trigger-workflows#workflow_run) that is supposed to allow you to configure a workflow to run based on the results of other workflows. I could not get that to work in my testing. The value of that would be setting the quick tests and unit tests to only run when the sniffs and ruleset checks complete successfully.

While it would be nice to set up this dependency, I think we could merge regardless in the interest of getting tests running again.

I also noticed that the `nightly` build was running with PHPCS 3.5.0, [resulting in a failure](https://travis-ci.com/github/WordPress/WordPress-Coding-Standards/jobs/454312331). I updated this to use 3.5.7 instead.

### CODE REVIEW

**Comment by dingo-d** [comment]
I've started to work on something similar for wptt sniffs. Still need to tackle the tests. I've decided to split tests and sniff part into separate workflows: https://github.com/WPTT/WPThemeReview/pull/261

The tricky part that is missing in GH actions is that there are still no allowed failures. Not sure how that will impact the tests.

**Comment by jrfnl** [comment]
> The tricky part that is missing in GH actions is that there are still no allowed failures.

@dingo-d From what I've seen, there _is_ (well, in a way), just not in the scripting part.

With Travis you had *one* status per complete run reporting and that status is added to required statusses to pass before a PR can be merged, which is why "allowed failures" is important.

With GH actions - you get individual reports in a PR on all the different build steps and instead, for those "builds" which are allowed to fail, you don't add a checkmark next to them in the "required statusses" for PRs to be merged.

Does that make sense ? or am I overlooking something ?

@desrosj One thing I don't know: while the PR has not been merged yet - how can we see what the current status would be and whether it works ?
I looked at the [Actions](https://github.com/WordPress/WordPress-Coding-Standards/actions/) tab here in the repo, but nothing is showing up yet.

**Comment by dingo-d** [comment]
> how can we see what the current status would be and whether it works?

I use https://github.com/nektos/act/ top locally test the actions. Works ok with this [instruction](https://github.com/shivammathur/setup-php#local-testing-setup).

The actions need to be in the `.github/workflows` folder for them to run, and it looks like @desrosj put the test yaml file in `.github/sniffs.yml/tests.yml` 🤷🏼‍♂️  Could be a typo.

@jrfnl regarding the allowed failures, there is a continue-on-error, but I don't think that's the same. At least not from what I've read on the [support question](https://github.com/actions/toolkit/issues/399).

**Comment by desrosj** [comment]
Sorry, I ran out of time to finish this week. This is definitely near the top of my list though.

The directory is indeed a typo. I'll get it all wrapped up early next week, and we can get everything moved over!

**Comment by dingo-d** [comment]
I'm making some progress on WPThemeReview implementation: https://github.com/WPTT/WPThemeReview/pull/261/checks

There are a couple of things I need to do:

 - I need to figure out how to add the checks so that only the highest/lowest PHPCS/WPCS combinations are used.
   For instance, [this check](https://github.com/WPTT/WPThemeReview/pull/261/checks?check_run_id=1542072811) passes, and [this](https://github.com/WPTT/WPThemeReview/pull/261/checks?check_run_id=1542072829) fails, but we never actually test the lowest supported PHPCS version with the develop version of WPCS. Also, I raised the versions compared to [here](https://github.com/WPTT/WPThemeReview/pull/259/checks) so that could cause breakage.
- I need to separate PHP 5.4 and 5.5 versions into a separate job, because they won't build on `ubuntu-latest` if [this table](https://github.com/shivammathur/setup-php#github-hosted-runners) is taken into the consideration.
- I plan on separating 'nightly' (8.1) into its own job that is allowed to fail. I may add it to the original, and do some checks in the run command for this once I play with it a bit
- Sniff stage fails because the `diff -B --tabsize=4 ./WPThemeReview/ruleset.xml <(xmllint --format "./WPThemeReview/ruleset.xml")` is showing differences and I have no idea why (this passed on travis) 🤷🏼‍♂️

**Comment by desrosj** [comment]
Alright, I think this is good for a review! I updated the PR description with more details to call out.

One thing that I did not include here is an update to the README file with new badges. Those can be grabbed from the workflow pages. I am happy to do that in a follow up PR, or I can include them here.

File: .github/workflows/ruleset.yml
Comment (dingo-d) [inline .github/workflows/ruleset.yml:24]: This will always be just 7.4, or was this added here for a future use case if we want to add different PHP versions?

**Comment by dingo-d** [comment]
Wow, amazing job! I'll definitely use some of this in WPThemeReview 😄  I've left one question regarding matrix usage.

The only thing that I'm worried about is that we get a failed test (which is allowed to fail), but we don't really have any easy way to distinguish allowed vs not allowed failures. I hope GH fixes this quickly.

File: .github/workflows/ruleset.yml
Comment (desrosj) [reply .github/workflows/ruleset.yml:24]: I don't think there was any particular reason. We could remove it in favor of hard coding the version below, or leave it in case we add other versions for testing. It probably makes it more clear at the very beginning of the job which versions are used for testing. But no strong preference either way.

**Comment by desrosj** [comment]
> The only thing that I'm worried about is that we get a failed test (which is allowed to fail), but we don't really have any easy way to distinguish allowed vs not allowed failures.

Hmm, we _could_ add a prefix to the job names that are allowed to fail. Maybe something like changing `PHP 7.4 on PHPCS 4.0.x-dev as 3.9.99` to `(Dev) PHP 7.4 on PHPCS 4.0.x-dev as 3.9.99`.

**Review by jrfnl — CHANGES_REQUESTED** [review]
Hi @desrosj Thanks for taking on this task! Awesome job with a complex script.

I've left some suggestions and questions inline.

Additionally:
* The `.gitattributes` file needs to be updated (removing `.travis.yml`)
* The CI badges in the readme need to be updated. As there are now several workflows, this may need a think through of how we want to present the status.
* A search should be done of the repo to see if there are other references to Travis (possibly in the `contributing` docs?) and if found, these should be removed/texts rewritten.

Other than that (not your concern):
- [ ] The required statuses for the branch protection will need to be updated for this to be mergable.
    This should only be done once the job names have been finalized.
- [ ] All open PRs will need to be rebased once this PRs has been merged.

File: .github/workflows/quicktest.yml
Comment (jrfnl) [reply .github/workflows/quicktest.yml:23]: Scratch that second point - I just noticed that the workflow name is prefixed to the job name by default if there are multiple workflows. So this would only apply to single workflow repos and multi-workflow with multiple jobs in a single workflow.

Comment (jrfnl) [reply .github/workflows/quicktest.yml:23]: Actually, in the "required status" overview, the workflow name is not prefixed, so it would be impossible to tell the difference between the `quicktest` and the `test` workflow when setting the statuses, so my original remark stands.

File: .github/workflows/quicktest.yml
Comment (jrfnl) [reply .github/workflows/quicktest.yml:27]: Pro-tip: `7.4` can actually probably better be changed to `latest` for this job. `latest` will always translate to the last stable release of PHP.

File: .github/workflows/quicktest.yml
Comment (desrosj) [reply .github/workflows/quicktest.yml:23]: Ah, thanks for catching this! I originally had a `lint` key in the matrix, but then I removed it because linting should always run for the quick test and unit test workflows, so it was unnecessary.

File: .github/workflows/quicktest.yml
Comment (desrosj) [reply .github/workflows/quicktest.yml]: Updated! I had left it as 7.4 as that is what the TravisCI file had and I was aiming for a apples to apples conversion. But this comment documents that, so it now uses `latest`. I usually default to using an explicit version to prevent unintended changes, but `latest` works for me if that's the preference.

File: .github/workflows/quicktest.yml
Comment (desrosj) [reply .github/workflows/quicktest.yml:57]: I looked into this a little bit. It seems based on [this](https://github.com/shivammathur/setup-php/blob/7b44f7e007410d13d3ed2deec1d452256eb79d96/src/install.ts#L27) and [this](https://github.com/shivammathur/setup-php/blob/7b44f7e007410d13d3ed2deec1d452256eb79d96/src/utils.ts#L20-L42) that an empty string would be returned if the input is not required, and a value is not supplied.

But it's probably better for clarity to just define `none`. It will make it more clear for someone unfamiliar with `setup-php`.

File: .github/workflows/quicktest.yml
Comment (desrosj) [reply .github/workflows/quicktest.yml:28]: Looks like using PHP 8.0 was causing the [workflow to fail](https://github.com/WordPress/WordPress-Coding-Standards/runs/1597986278?check_suite_focus=true). I've reverted it for now in the interest of getting the automated tests fixed and we can revisit this after.

File: .github/workflows/quicktest.yml
Comment (desrosj) [reply .github/workflows/quicktest.yml:63]: Makes sense to me. Are you thinking on the workflow level? Or a step that exports the desired value to an environment variable?

File: .github/workflows/ruleset.yml
Comment (desrosj) [reply .github/workflows/ruleset.yml:1]: I defer to you as a maintainer for that. I go back and forth myself. But, I think personally, I like having separate workflows if only for the reason that it makes it easier to see what has failed and what has passed. You don't have to click into the workflow to see which job failed.

File: .github/workflows/ruleset.yml
Comment (desrosj) [reply .github/workflows/ruleset.yml:64]: Yep, those can be combined if that's the preference. My thinking for breaking them into separate steps is you'll know exactly which check failed without having to expand the results.

File: .github/workflows/ruleset.yml
Comment (desrosj) [reply .github/workflows/ruleset.yml:82]: I think I understand, but not positive. Can you suggest the change?

File: .github/workflows/quicktest.yml
Comment (desrosj) [reply .github/workflows/quicktest.yml:1]: Sure! Would QTest work for consistency with the job prefix?

File: .github/workflows/quicktest.yml
Comment (jrfnl) [reply .github/workflows/quicktest.yml:63]: This is how I did it myself in other, similar repos:

https://github.com/PHPCSStandards/PHPCSUtils/blob/3f1a79a35256b41ac5481ab743d04c144365db30/.github/workflows/quicktest.yml#L38-L54

File: .github/workflows/ruleset.yml
Comment (jrfnl) [reply .github/workflows/ruleset.yml:64]: The thing is:
1. The step which failed will be expanded automatically by GH anyway.
2. You'd want to expand that step either way as you need to see exactly on _what_ (line) it failed.

File: .github/workflows/ruleset.yml
Comment (jrfnl) [reply .github/workflows/ruleset.yml:1]: In my thinking:
1. The jobs will be listed separately in the "Some checks were not successful" list below anyway.
2. The jobs can very easily be viewed from the PR -> Checks tab, like here: https://github.com/WordPress/WordPress-Coding-Standards/pull/1965/checks
3. I like DRY ;-)

@dingo-d @GaryJones What's your opinion ?

Comment (desrosj) [reply .github/workflows/ruleset.yml:1]: Thinking it through doing the dishes just now, the other advantage of keeping them separate is that if the conditions need to change in the future, it's easier to change them for just one of the jobs while leaving the other the same.

File: .github/workflows/ruleset.yml
Comment (jrfnl) [reply .github/workflows/ruleset.yml:82]: I'm not 100% fluent in bash as I work on Windows myself, but I think this example should hopefully give you enough to go on:
https://github.com/squizlabs/PHP_CodeSniffer/issues/1818#issuecomment-592951176

File: .github/workflows/ruleset.yml
Comment (jrfnl) [reply .github/workflows/ruleset.yml:1]: A hypothetical future change, to me, is not a reason not to use DRY now.

Comment (dingo-d) [reply .github/workflows/ruleset.yml:1]: As the two are like 90% the same (as in the initial setups are the same), I'm fine with adding the sniffs workflow as a separate job here, reduces the boilerplate. Future changes should be easy to add using matrix includes and excludes and few environment variables 😄

File: .github/workflows/ruleset.yml
Comment (dingo-d) [reply .github/workflows/ruleset.yml:64]: I think it's easier to check for the failures if they are all separated like here. If they are in one step, the entire step will fail if just one fails, but I think that there could be a bit more mental overhead to checking which step exactly failed (was it core, or docs or extras, etc.) I'm fine with it being separated as is.

File: .github/workflows/ruleset.yml
Comment (dingo-d) [inline .github/workflows/ruleset.yml:62]: The `--no-suggest` should be removed here as well, no?

File: .github/workflows/ruleset.yml
Comment (desrosj) [reply .github/workflows/ruleset.yml:64]: Going to leave this one as is. We can consolidate later if we find separating is unnecessary.

File: .github/workflows/quicktest.yml
Comment (desrosj) [reply .github/workflows/quicktest.yml]: I think I have configured this properly and it seems to work, however it's not clear if there are any potential conflicts with running requiring a specific version of PHPCS in the step before and then the action running `composer install` after.

**Comment by desrosj** [comment]
I went through this PR again, and I think that all of the feedback has been addressed with the exception of the workflow names. I'm happy to name them whatever you think would be best, though! Feel free to suggest what you'd like to see them be.

**Review by dingo-d — COMMENTED** [review]
Looks good to me, also I don't think I have anything against naming, so it's ok on that front from me 🙂

**Review by dingo-d — APPROVED** [review]

**Comment by jrfnl** [comment]
FYI: I did another review and found a number of things which would still need to be updated.

As this PR is a blocker for all other PR/progress in this repo, leaving it open any longer is not desirable.
With that in mind, @desrosj and me had a call this afternoon to talk through the review and the additional changes I was suggesting.

I'll be adding a number of additional commits to this PR based on our discussion to touch up and finalize this PR.

@dingo-d @GaryJones Please **squash** all commits on merge of this PR so as to keep the repo history semi-clean.

**Review by jrfnl — APPROVED** [review]
With the additional changes now pushed after the call/discussion with @desrosj, IMO this PR is ready to be merged.

**Comment by jrfnl** [comment]
@dingo-d @GaryJones I've just updated the protected branches. The status checks from Travis have been removed and the new ones from the GH Actions workflows have been added.

File: .github/workflows/ruleset-checks-sniffs.yml
Comment (dingo-d) [inline .github/workflows/ruleset-checks-sniffs.yml:60]:
```suggestion
        run: vendor/bin/phpcs --runtime-set ignore_warnings_on_exit 1
```

Extra space was here on purpose or a typo?

File: .github/workflows/unit-tests.yml
Comment (dingo-d) [inline .github/workflows/unit-tests.yml:85]:
```suggestion
        uses: ramsey/composer-install@v1
```

I think we can omit the quotes.
```suggestion
        uses: "ramsey/composer-install@v1"
```
```suggestion
        uses: "ramsey/composer-install@v1"
```
```suggestion
        uses: "ramsey/composer-install@v1"
```
```suggestion
        uses: "ramsey/composer-install@v1"
```
```suggestion
        uses: "ramsey/composer-install@v1"
```
```suggestion
        uses: "ramsey/composer-install@v1"
```
```suggestion
        uses: "ramsey/composer-install@v1"
```

File: .github/workflows/unit-tests.yml
Comment (dingo-d) [inline .github/workflows/unit-tests.yml:89]:
```suggestion
        uses: ramsey/composer-install@v1
```

Here as well

**Review by dingo-d — CHANGES_REQUESTED** [review]
Left few small comments, otherwise looks great. Awesome work @desrosj and @jrfnl !

**Comment by desrosj** [comment]
Thanks @dingo-d! Addressed those, I think we're good to go now!

File: .github/workflows/ruleset-checks-sniffs.yml
Comment (jrfnl) [reply .github/workflows/ruleset-checks-sniffs.yml]: Thanks for catching that. That was a typo for sure.

**Review by dingo-d — APPROVED** [review]

**Pull request merged**
